src/darwin/Framework/CHIP/MTRAttributeValueWaiter.mm

```diff
@@ -120,15 +137,45 @@ - (void)_notifyWithError:(NSError * _Nullable)error
         MTR_LOG("%@ %p wait for attribute values unknown error: %@", self, self, error);
     }
 
-    auto completion = self.completion;
-    auto queue = self.queue;
-    self.completion = nil;
-    self.queue = nil;
+    auto * device = self.device;
     dispatch_async(queue, ^{
+        // Call into the device from our async dispatch, to avoid problems when
+        // _notifyWithError is happening while the device lock is held.
+        //
+        // But this seems to trigger TSAN errors about null-derefs.  Probably
```

Comment:
Yeah you can't dispatch away in code that's called from `dealloc`. Sadly the runtime doesn't catch it when dealloc has attempted to "revive" the object and you just end up with UAFs.

Reply:
I think the best thing to do here is to not call `_forget...` from `_notifyWithError`. Instead leave it up to the caller.
- When MTRDevice calls `_notifyWithError`, remove the waiter explicitly first
- When MTRDevice cancels all waiters, make it call a new `_notifyOfCancellation` method on the waiter that just calls `_notifyWithError` but does not call back into MTRDevice
- Only `cancel` (or dealloc) called directly on the waiter needs to call `_forgetWaiter` on the device along with calling `_notifyOfCancellation`.

There's also the option to rely the weak NSHashTable to do some of its own cleanup. For example from `dealloc` we could simply call `_notifyOfCancellation` because we know there's nobody else holding the waiter alive, so the slot in the hash table is reclaimable the next time it's needed without doing explicit work.

For explicit `cancel` and for successful notifications from the MTRDevice it probably makes sense to explicitly forget the waiter; that way the table won't end up growing wildly even if the client is for some reason (bug?) hanging on to waiters that have already notified.